src/llama_recipes/utils/train_utils.py

```diff
@@ -87,6 +87,9 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
                     # if fp16 is enabled, use gradient scaler to handle gradient update
                     scaler.scale(loss).backward()
                     if (step + 1) % gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
+                        if train_config.gradient_clipping > 0.0:
+                            scaler.unscale_(optimizer)
+                            torch.nn.utils.clip_grad_norm_(model.parameters(), train_config.gradient_clipping)
```

Comment:
@gpengzhi thanks for the PR, in case FSDP we would need to use its own [clip_grad](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1067) as its shard aware. It would be `model.clip_grad_norm_(clip_threshold)`. Can you pls modify accordingly to cover FSDP as well.

Reply:
Thank you for the helpful feedback! 

Refactored the code to take care of the FSDP scenario.